server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -912,6 +913,7 @@ public void perform(Request request, ActionListener<PrimaryResult<ReplicaRequest
                     return result;
                 });
             }
+            assert indexShard.getActiveOperationsCount() != 0 : "must perform shard operation under a permit";
```

Comment:
\> 0?

Reply:
We can't, because some actions (e.g. `TransportVerifyShardBeforeCloseAction`) run under an operation block rather than a single permit.

Reply:
Makes sense, thanks.